File: mobile_order/lib/models/order.dart
//file: \lib\models\order.dart

class SelectedOption {
  final String name; // 옵션 이름 (예: "샷 추가")
  final String choice; // 사용자가 선택한 옵션 값 (예: "추가")
  final int? price; // 선택지에 따른 추가 가격
  final int? quantity; // 선택지에 따른 수량

  SelectedOption({
    required this.name,
    required this.choice,
    this.price, // 가격을 int로 설정
    this.quantity, // 수량은 nullable
  });

  factory SelectedOption.fromJson(Map<String, dynamic> json) {
    return SelectedOption(
      name: json['name'],
      choice: json['choice'],
      price: json['price'] as int?, // int로 변환
      quantity: json['quantity'] as int?, // nullable로 변환
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'choice': choice,
      'price': price,
      'quantity': quantity,
    };
  }
}

class OrderItem {
  final String id;
  final String name;
  final int price;
  final int quantity;
  final bool isComplimentary;
  final List<SelectedOption> selectedOptions; // 선택된 옵션 목록

  OrderItem({
    required this.id,
    required this.name,
    required this.price,
    required this.quantity,
    this.isComplimentary = false,
    this.selectedOptions = const [], // 기본값으로 빈 목록
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'],
      name: json['name'],
      price: json['price'] as int,
      quantity: json['quantity'],
      isComplimentary: json['isComplimentary'] ?? false,
      selectedOptions: (json['selectedOptions'] as List<dynamic>?)
              ?.map((option) =>
                  SelectedOption.fromJson(option as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'price': price,
      'quantity': quantity,
      'isComplimentary': isComplimentary,
      'selectedOptions':
          selectedOptions.map((option) => option.toJson()).toList(),
    };
  }
}

class Order {
  final String? id;
  final String restaurantId;
  final String businessDayId; // 필수 필드로 변경
  final int tableId;
  final List<OrderItem> items;
  final String status;
  final int totalAmount;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isComplimentaryOrder;
  final String user;

  Order({
    this.id,
    required this.restaurantId,
    required this.businessDayId, // 필수 매개변수로 변경
    required this.tableId,
    required this.items,
    required this.status,
    required this.totalAmount,
    required this.createdAt,
    required this.updatedAt,
    this.isComplimentaryOrder = false,
    required this.user,
  });

  Order copyWith({
    String? id,
    String? restaurantId,
    String? businessDayId,
    int? tableId,
    List<OrderItem>? items,
    String? status,
    int? totalAmount,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isComplimentaryOrder,
    String? user,
  }) {
    return Order(
      id: id ?? this.id,
      restaurantId: restaurantId ?? this.restaurantId,
      businessDayId: businessDayId ?? this.businessDayId,
      tableId: tableId ?? this.tableId,
      items: items ?? this.items,
      status: status ?? this.status,
      totalAmount: totalAmount ?? this.totalAmount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isComplimentaryOrder: isComplimentaryOrder ?? this.isComplimentaryOrder,
      user: user ?? this.user,
    );
  }

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['_id'],
      restaurantId: json['restaurantId'],
      businessDayId: json['businessDayId'],
      tableId: json['tableId'],
      items: (json['items'] as List)
          .map((item) => OrderItem.fromJson(item))
          .toList(),
      status: json['status'],
      totalAmount: json['totalAmount'],
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      isComplimentaryOrder: json['isComplimentaryOrder'] ?? false,
      user: json['user'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'restaurantId': restaurantId,
      'businessDayId': businessDayId,
      'tableId': tableId,
      'items': items.map((item) => item.toJson()).toList(),
      'status': status,
      'totalAmount': totalAmount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'isComplimentaryOrder': isComplimentaryOrder,
      'user': user,
    };
  }
}

File: server/models/Order.js
//file: \server\models\Order.js

const mongoose = require("mongoose");
// 옵션 스키마
// 옵션 스키마
const SelectedOptionSchema = new mongoose.Schema({
  name: { type: String, required: true }, // 옵션 이름 (예: "샷 추가")
  choice: { type: String, required: true }, // 사용자가 선택한 옵션 값 (예: "추가")
  price: { type: Number, default: 0 }, // 선택지에 따른 추가 가격 (없을 경우 0)
  quantity: { type: Number }, // 선택지에 따른 수량 (없을 수 있음)
});

const OrderItemSchema = new mongoose.Schema({
  id: { type: String, required: true },
  name: { type: String, required: true },
  price: { type: Number, required: true },
  quantity: { type: Number, required: true },
  selectedOptions: {
    type: [SelectedOptionSchema], // 옵션 배열
    default: [], // 옵션이 없을 경우 기본값으로 빈 배열
  },
});

const OrderSchema = new mongoose.Schema({
  restaurantId: { type: String, required: true },
  businessDayId: { type: mongoose.Schema.Types.ObjectId, ref: "BusinessDay", required: true }, // 추가된 부분
  tableId: { type: Number, required: true },
  items: [OrderItemSchema],
  status: {
    type: String,
    enum: ["pending", "preparing", "served", "completed"],
    default: "pending",
  },
  isComplimentaryOrder: { type: Boolean, default: false },
  totalAmount: { type: Number, required: true },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
  user: { type: String, required: true }, // ObjectId에서 String으로 변경
});

// pre-save 훅 수정 (businessDayId 관련 로직 추가)
OrderSchema.pre("save", async function (next) {
  if (this.isNew) {
    const Restaurant = mongoose.model("Restaurant");
    const BusinessDay = mongoose.model("BusinessDay");

    const restaurant = await Restaurant.findOne({ restaurantId: this.restaurantId });

    if (!restaurant) {
      throw new Error("Restaurant not found");
    }

    if (!restaurant.hasTables) {
      throw new Error("This order model is for restaurants with tables only");
    }

    // businessDayId가 유효한지 확인
    const businessDay = await BusinessDay.findById(this.businessDayId);
    if (!businessDay) {
      throw new Error("Business day not found");
    }

    // 필요한 경우 여기에 추가 로직을 구현할 수 있습니다.
    // 예: 주문 번호 생성, BusinessDay 모델 업데이트 등
  }
  next();
});

module.exports = mongoose.model("Order", OrderSchema);
